### PR TITLE
custom processor for omni poc

### PR DESCRIPTION
**Pull request by subawocit**
# Description

This PR introduces the implementations for hybrid multimodal processor in MaxText.

Specifically, it implements modular hybrid multimodal support in the preprocessing and fusion pipelines:
- **Text Prompt**: Formatted with the text decoder (Qwen 3) chat template and special vision tags (`<|vision_start|>`, `<|image_pad|>`, `<|vision_end|>`) and tokenized by the text decoder's tokenizer.
- **Image Inputs**: Preprocessed by the vision encoder (Gemma 3) processor.
- **Fusion Stage**: Handled by `processor_maxtext_omni.py`, which expands placeholder tokens into 256 copies of `<|image_pad|>`. This ensures the text decoder knows precisely where to inject the 256 projected visual embeddings within the token stream without corrupting text subword tokenization or array indexing.

This is Step 3 of a 5-step Proof-of-Concept for any-to-any multimodal alignment in MaxText. The overall goal is to align pretrained vision encoders with text-only LLMs by connecting the two with a dynamically configured adaptation layer (customized MLP connector), special tokenizer mapping for visual placeholder tokens, and training only the MLP using supervised fine tuning.

# Files

1. **Omni Multimodal Processor** 
   - `experimental/omni_poc/utils/processor_maxtext_omni.py`
     - Implemented modular visual placeholder token management (`<|vision_start|>`, `<|image_pad|>`, `<|vision_end|>`) combining Qwen 3's vocabulary with Gemma 3's visual token count (256 tokens per image).
     - Decoupled architectures via `VISION_TOKENS_PER_IMAGE` and `DECODER_SPECIAL_TOKENS` lookup tables.
     - Added functions for offset calculation, placeholder token expansion, and bidirectional mask generation (`get_image_offsets_omni`, `add_extra_tokens_for_omni`, `get_bidirectional_mask_vision_omni`) with validation for unsupported combinations.

2. **Name Registration & Routing** 
   - `multimodal/processor.py`, `layers/decoders.py`, and `layers/nnx_decoders.py`
     - Integrated stitched architecture routing into `processor.py` for models matching `"maxtext-omni"` in `model_name`.
     - Registered `maxtext-omni-gemma3-qwen3` model name across Flax/NNX decoders and multimodal registries.

3. **Preparation & Decoding Utilities** 
   - `experimental/omni_poc/prepare_checkpoint.sh`
     - Automated end-to-end HuggingFace download, conversion (`to_maxtext`), and stitching workflow.
   - `experimental/omni_poc/utils/decode_omni.py`
     - Added a decoding script to run prefill and autoregressive generation on ChartQA multimodal validation samples.
     - Update after review:
       - Replaced the repeated full-sequence forward loop with a decoupled prefill step and Flax KV-cached autoregressive decoding, reducing vision encoder passes from O(N x V_ViT) to O(1 x V_ViT) and attention complexity per generated token from quadratic O((L + i)^2) to linear O(L + i).
       - Replaced the Python step loop and host-device synchronization barriers with an XLA-compiled jax.lax.scan generation kernel, reducing CPU-TPU roundtrips from O(N) blocking calls to O(1) pure on-chip execution.
       - L = Prefill / Prompt length; N = Number of generated tokens; V_ViT = Cost of a full Gemma 3 Vision Transformer forward pass
       - In total, this reduced the decoding time from 3.3 minutes to 6 seconds in one example.
   - `experimental/omni_poc/utils/stitch_checkpoint.py`
     - Modified the checkpoint stitch script to be compatible with current codebase.

4. **Configuration** 
   - `experimental/omni_poc/maxtext-omni-gemma3-qwen3.yml`
     - Model configuration defining Gemma 3 vision tower parameters, Qwen 3 4B LLM backbone parameters, and custom projector.

5. **Unit Tests** 
   - `experimental/omni_poc/tests/processor_maxtext_omni_test.py`
     - Added unit tests for Omni visual token expansion, offset calculations, dtype preservation, and combination validation.
   - `experimental/omni_poc/tests/stitch_checkpoint_test.py`
     - Unit tests for checkpoint stitching compatibility.
   - `tests/unit/multimodal_utils_test.py`
     - Added unit tests for `maxtext-omni-gemma3-qwen3` routing, fusion masking, and exception handling in the multimodal processor.

# Tests

## 1. Checkpoint Stitching Generation
Generated a stitched Orbax checkpoint from converted Gemma 3 4B vision weights and Qwen 3 4B language weights:
```bash
JAX_PLATFORMS=cpu python -m maxtext.experimental.omni_poc.utils.stitch_checkpoint \
    src/maxtext/experimental/omni_poc/maxtext-omni-gemma3-qwen3.yml \
    --vision_load_path=gs://yuchenhou-maxtext-logs/omni_checkpoints/gemma3-4b_converted/0/items \
    --llm_load_path=gs://yuchenhou-maxtext-logs/omni_checkpoints/qwen3-4b_converted/0/items \
    --stitched_output_path=gs://yuchenhou-maxtext-logs/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items
```

## 2. Checkpoint Stitching Unit Test
Verified that the stitched checkpoint accurately preserves original weights from both source models:
```bash
OMNI_TEST_BASE_DIR='gs://yuchenhou-maxtext-logs/omni_checkpoints' JAX_PLATFORMS=cpu \
    /home/yuchenhou_google_com/maxtext_env/bin/python -m unittest src/maxtext/experimental/omni_poc/tests/stitch_checkpoint_test.py
```
**Output:**
```
----------------------------------------------------------------------
Ran 2 tests in 135.804s

OK
```

## 3. Omni Processor Unit Test
Verified token ID mappings, sequence expansions, and offsets:
```bash
python -m unittest src/maxtext/experimental/omni_poc/tests/processor_maxtext_omni_test.py
```
**Output:**
```
----------------------------------------------------------------------
Ran 3 tests in 0.001s

OK
```

## 4. Multimodal Processor Unit Tests
Verified hybrid multimodal routing and fusion:
```bash
JAX_PLATFORMS=cpu python3 -m pytest -v tests/unit/multimodal_utils_test.py tests/unit/qwen3_omni_layers_test.py
```
**Output:**
```
============================= 56 passed, 1 skipped in 192.32s =============================
```

## 5. End-to-End Multimodal Decoding
Ran end-to-end decoding on ChartQA validation samples using the stitched checkpoint:
```bash
JAX_PLATFORMS=tpu /home/yuchenhou_google_com/maxtext_env/bin/python src/maxtext/experimental/omni_poc/utils/decode_omni.py \
    --checkpoint_path="gs://yuchenhou-maxtext-logs/omni_checkpoints/omni_stitched_gemma3-4b_qwen3-4b/0/items" \
    --num_samples=2 \
    --max_new_tokens=64
```
**Output:**
```text
Sample 1 (Index 1309):
  Question: What was the average annual total income per household of those in the top decile group?
  Ground Truth: ['186600']
  Model Response: <think>
Okay, the user is asking about the average annual total income per household in the top decile group. First, I need to figure out what data they're referring to. The question is pretty general, so I should consider different sources where this information might be available.

Sample 2 (Index 228):
  Question: Which gender is represented by the red color line?
  Ground Truth: ['Male']
  Model Response: <think>
Okay, the user is asking about the gender represented by the red color line. But wait, the initial message they provided is a mix of different languages and some random characters. It seems like they might have pasted some text that's not properly formatted or maybe a mix of different scripts.

```

This ensures the data is flowing end to end. The response itself does not make sense due to the randomly initialized projector, as expected.

# Step-by-Step Objectives
- [PR 4485](https://github.com/google/maxtext/pull/4485) Multi-Directory Checkpoint Restoration: Implemented selective sub-tree parameter loading using Orbax such that we can initiaze a multimodal model with parameters from multiple checkpoints.
- [PR 4593](https://github.com/google/maxtext/pull/4593) Dynamic MLP Connector: Add omni modal adapter layer (MLP) to connect vision tower output to the LLM decoder
- [PR 4746](https://github.com/google/maxtext/pull/4746) and [PR 4839](https://github.com/google/maxtext/pull/4839) Special Tokenizer & Placeholder Masking: Add special token <|image|> to tokenizer and make sure the masking is correct in the decoder.
- [PR 4864](https://github.com/AI-Hypercomputer/maxtext/pull/4864) COCO-Narratives Data Pipeline Processing
- SFT Execution & Evaluation

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -3853,6 +3854,7 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
           "qwen3-vl-30b-a3b",
           "qwen3.5-35b-a3b",
           "qwen3.5-397b-a17b",
+          "omni-gemma3-qwen3",
```
Comment (hengtaoguo) [inline src/maxtext/configs/types.py]: Thanks for the addition! I would still prefer to name this `maxtext-omni-poc`. Let's chat about that more.

File: src/maxtext/experimental/omni_poc/omni-gemma3-qwen3.yml [hunk 2]
```diff
@@ -34,6 +44,7 @@ logits_via_embedding: true
 normalize_embedding_logits: false
 enable_dropout: false
 tokenizer_type: "huggingface"
+tokenizer_path: "Qwen/Qwen2.5-7B-Instruct"
```
Comment (hengtaoguo) [inline src/maxtext/experimental/omni_poc/omni-gemma3-qwen3.yml]: Is this the right tokenizer?

File: src/maxtext/multimodal/processor.py [hunk 3]
```diff
@@ -138,8 +140,13 @@ def preprocess_image_for_training(image, config):
 def get_image_offsets(config, processor_output: mm_utils.PreprocessorOutput | None):
   """Get the increase in total token count after inserting image token placeholders"""
   vision_block = _get_vision_block(config)
+  decoder_block = _get_decoder_block(config)
 
-  if vision_block in ["gemma3"]:
+  if vision_block == "gemma3" and decoder_block == "qwen3":
```
Comment (hengtaoguo) [inline src/maxtext/multimodal/processor.py]: I wonder if there's a more elegant way registering maxtext-omni preprocessing func, without specifically calling out the block name. (At least we should add a comment string for this if branch saying `# This is for stitched model`)

Maybe check if `vision_block==decoder_block`? If not, we could add a dictionary `decoder_block: preprocess_func` above and use such function based on the decoder key. Let's sync offline.

File: src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py
Comment (hengtaoguo) [inline src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py:1]: For this file, if I understand correctly, it:

(1) Compute number of visual tokens based on `vision_block`.
(2) Adapt to text-only tokenizer (which should be specific to `decoder_block`?)

I wonder if we could do a router-pattern: based on vision_block and decoder_block, it triggers different branches. For this PR, we can just enable gemma3/qwen3 combination, but leave other branches raise error "stitched model not supported for XXX". Let sync shortly?

File: src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  """Calculate the increase in total token count after inserting visual token sequences."""
+  has_images = processor_output is not None and processor_output.pixel_values is not None
+  num_images = processor_output.pixel_values.shape[0] if has_images else 1
+  # +256 for <|image_pad|>, -1 for original placeholder (<|vision_start|> and <|vision_end|> already in prompt)
+  return (DEFAULT_NUM_TOKENS_PER_IMAGE - 1) * num_images
+
+
+def add_extra_tokens_for_omni(
+    tokens,
+    config=None,
```
Comment (hengtaoguo) [inline src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py]: `config` is not used in this function. Is it because the upper stream requires such an input?

**Review by hengtaoguo — COMMENTED** [review]
Thanks for the PoC PR! Let's sync offline to discuss the design pattern. We should make the stitched model as intuitive and robust as possible.

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (subawocit) [reply src/maxtext/configs/types.py]: Update: rename models to `maxtext-omni-{vision encoder's model name}-{text decoder's model name}`

File: src/maxtext/experimental/omni_poc/omni-gemma3-qwen3.yml (hunk 2 above)
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/omni-gemma3-qwen3.yml]: Apologies! I've updated it to the correct tokenizer path and re-run everything from `stitching->processing->sft->evaluation`. All results were compared between before-fix and after-fix, no errors detected

File: src/maxtext/multimodal/processor.py (hunk 3 above)
Comment (subawocit) [reply src/maxtext/multimodal/processor.py]: Updated - now the processor checks if the model name contains `maxtext-omni`, and if so, routes to our custom processor in `experimental/omni_poc/utils/processor_maxtext_omni.py`

File: src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py:1]: Updated - now the custom processor checks `vision_block` and `decoder_block` and raise errors if the combination isnt supported

File: src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py (hunk 4 above)
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/utils/processor_omni_gemma3_qwen3.py]: We can determine which processor to use solely based on `vision_block` and `decoder_block`, so config isn't necessary here. However, if future implementations require config, we can replace the current input with the config and recompute the vision and decoder block names inside the custom processor. I'll leave a note in my handover doc

File: src/maxtext/experimental/omni_poc/utils/processor_maxtext_omni.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    expanded_tokens = []
+    for token in token_list:
+      if token == image_pad_id:
+        expanded_tokens.extend([image_pad_id] * num_tokens_per_image)
+      else:
+        expanded_tokens.append(token)
+    return np.array(expanded_tokens, dtype=dtype)
+
+  # Case 2: No placeholder in text, but image data is present, prepend vision block
+  if processor_output is not None and processor_output.pixel_values is not None:
```
[the remark this answers is not in the document]
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/utils/processor_maxtext_omni.py:92]: Case 2 is only a safety fallback for manual scripts/tests that **skip reformat_prompt**. But let me add a warning

File: src/maxtext/configs/types.py [hunk 6]
```diff
@@ -248,6 +248,7 @@ class ProfilerType(str, Enum):
     "gemma2-27b",
     "gemma3-4b",
     "gemma3-12b",
+    "maxtext-omni-gemma3-qwen3",
```
Comment (hengtaoguo) [inline src/maxtext/configs/types.py]: Maybe move this line under `gemma4-e4b`.

**Review by hengtaoguo — APPROVED** [review]
Thanks for the great work!

File: src/maxtext/configs/types.py (hunk 6 above)
Comment (subawocit) [reply src/maxtext/configs/types.py]: Got it!

File: src/maxtext/experimental/omni_poc/utils/decode_omni.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+  positions = np.tile(np.arange(seq_len, dtype=np.int32), (1, 1))
+  segment_ids = np.zeros((1, seq_len), dtype=np.int32)
+  segment_ids[:, :true_length] = 1
+
+  # Autoregressive decoding loop
+  generated_tokens = []
+  curr_len = true_length
+
+  with jax.set_mesh(mesh):
+    # Initial prefill pass
```
Comment (lydhr) [inline src/maxtext/experimental/omni_poc/utils/decode_omni.py]: In my local test, the decode_omni took approximately 5.5 minutes to decode a single sample. It is likely due to the missing KV caching in the prefill pass inside decoding loop.
--> _forward_step

File: src/maxtext/experimental/omni_poc/utils/decode_omni.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+def _forward_step(model, params, tokens, positions, segment_ids, images):
+  """Executes a single forward pass of the model and returns output logits."""
+  return model.apply(
+      {"params": params},
+      decoder_input_tokens=tokens,
+      decoder_positions=positions,
+      decoder_segment_ids=segment_ids,
+      encoder_images=images,
+      enable_dropout=False,
+      model_mode="prefill",
```
Comment (lydhr) [inline src/maxtext/experimental/omni_poc/utils/decode_omni.py:177]: When using model_mode="prefill", (prompt + images + generated tokens so far) are re-compute for every single token generated.

**Review by lydhr — APPROVED** [review]
LGTM. Feel free to wrap up this PR and consider optimizing KV caching in decoding in future PRs.

File: src/maxtext/experimental/omni_poc/utils/decode_omni.py (hunk 7 above)
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/utils/decode_omni.py]: Good catch! I've updated my code to optimize the performance, see below.

File: src/maxtext/experimental/omni_poc/utils/decode_omni.py (hunk 8 above)
Comment (subawocit) [reply src/maxtext/experimental/omni_poc/utils/decode_omni.py:177]: I replaced the repeated full-sequence forward loop with a decoupled prefill step and Flax KV-cached autoregressive decoding. This reduced the time from `Prefill: 11.00s | Generation: 197.45s (64 tokens, 0.3 tok/s)` to `Prefill: 3.80s | Generation: 6.18s (64 tokens, 10.4 tok/s)`
